talos-bootstrap

```diff
@@ -6,9 +6,53 @@
 # This code should (try to) follow Google's Shell Style Guide
 # - https://google.github.io/styleguide/shell.xml
 #
-
 node=
 OP=
+
+# Prepend $HOME/.tbs/bin to $PATH so we pickup on any 
+# downloaded/preferred binary versions
+export PATH=$HOME/.tbs/bin:$PATH
+
+# Test for a binary, and prompt to install to /usr/local/bin, 
+# or download temp copy using J.Pillora's hosted copy of installer:
+# https://github.com/jpillora/installer
+function check_or_install {
+  PROG=$1
+  REPO=$2
+  # TODO: Versioned install/download
+  VERSION=${4:-"latest"}
+  METHOD=${3:-}
+  NEEDS_INSTALLED=
+  BIN_DIR=$HOME/.tbs/bin
+  # test if $PROG already exists: if yes NEEDS_INSTALLED remains unset,
+  # therefore we skip over the 
+  which $PROG > /dev/null || NEEDS_INSTALLED=true
+  # No install method specified, prompt
+  if [[ $NEEDS_INSTALLED && -z "$METHOD" ]]; then
+    install_option=$(dialog --keep-tite --title "Missing: $PROG..." \
+                      --menu "Required program missing: $PROG\nSelect an option" 0 0 0 \
+                      1 "Install $PROG to /usr/local/bin" \
+                      2 "Download temporary copy of $PROG" \
+                      3 "Do nothing and exit" 3>&1 1>&2 2>&3) || exit 0
+    case $install_option in
+      1) METHOD="install" ;;
+      2) METHOD="download" ;;
+      *) exit 0 ;;
+    esac
+  fi
+  if [[ $NEEDS_INSTALLED && $METHOD == "install" ]]; then
+    curl -s "https://i.jpillora.com/${REPO}\!?as=${PROG}&type=script" | bash
+  elif [[ $NEEDS_INSTALLED && $METHOD == "download" ]]; then
+    # store PWD
+    pushd $PWD > /dev/null
+    mkdir -p $HOME/.tbs/bin && cd $HOME/.tbs/bin
+    # url without ! will attempt to download to $PWD
+    curl -s "https://i.jpillora.com/${REPO}?as=${PROG}&type=script" | bash
```

Comment:
1. This adds additional dependency - `bash` which was not required to use this script before. 

2. I have a fear using external services without informing user about this. `curl | bash` is very bad pattern to include into script. But it can be described in a README.md

Reply:
How about parsing architecture and platform using `uname` and preform url from this:

```
https://github.com/siderolabs/talos/releases/latest/download/talosctl-<platform>-<arch>
```

Or if version specified:
```
https://github.com/siderolabs/talos/releases/download/<version>/talosctl-<platform>-<arch>
```

See approach from this file: 

https://github.com/deckhouse/deckhouse/blob/ab6faa3ad220f203117626a15052333bf84154dc/Makefile#L11-L46

Reply:
Those are great points.

- I agree the `curl | bash` is a less-than-ideal pattern, as well as an additional dependency along with all the binaries that the generated install script from J.Pillora's installer checks for [although I believe all the checks are for binaries expected on a POSIX-compliant system]. The design was a "quick-fix" I chose w/o considering the extra deps it added

I'll see what I can do to address your concerns by specializing the check_or_install function to our specific use case for talosctl


FWIW, I initially chose to design in this way anticipation of another feature contribution which will allow the user to select the to-be-installed talosctl version from a list as it would need to ensure `yq` was available for easily parsing Github API results, **BUT** I can still accomplish that and narrow dependencies to just the actual binaries that we need: talosctl and yq.

Ill hammer away and get back with you :)

Reply:
Thank you @TrooperT ❤️ 

I mean, that I don't against using `curl | bash` but it should be described in `README.md` insead of placing it into script.
So user will invoke it at their own risk, or they still have the opportunity to install it from their distribution.

From the other side script can check if required binary exists and can be running.